File: QtClient/JoplinQtClient/webapi.cpp
#include <stable.h>

#include "webapi.h"

using namespace jop;

WebApi::WebApi(const QString &baseUrl) {
	baseUrl_ = baseUrl;
	sessionId_ = "";
	connect(&manager_, SIGNAL(finished(QNetworkReply*)), this, SLOT(request_finished(QNetworkReply*)));
}

QString WebApi::baseUrl() const {
	return baseUrl_;
}

void WebApi::execRequest(HttpMethod method, const QString &path, const QUrlQuery &query, const QUrlQuery &data, const QString& tag) {
	QueuedRequest r;
	r.method = method;
	r.path = path;
	r.query = query;
	r.data = data;
	r.tag = tag;
	r.buffer = NULL;
	queuedRequests_ << r;

	processQueue();
}

void WebApi::post(const QString& path,const QUrlQuery& query, const QUrlQuery& data, const QString& tag) { execRequest(HttpMethod::POST, path, query, data, tag); }
void WebApi::get(const QString& path,const QUrlQuery& query, const QUrlQuery& data, const QString& tag) { execRequest(HttpMethod::GET, path, query, data, tag); }
void WebApi::put(const QString& path,const QUrlQuery& query, const QUrlQuery& data, const QString& tag) { execRequest(HttpMethod::PUT, path, query, data, tag); }
void WebApi::del(const QString &path, const QUrlQuery &query, const QUrlQuery &data, const QString &tag) { execRequest(HttpMethod::DEL, path, query, data, tag); }
void WebApi::patch(const QString &path, const QUrlQuery &query, const QUrlQuery &data, const QString &tag) { execRequest(HttpMethod::PATCH, path, query, data, tag); }

void WebApi::setSessionId(const QString &v) {
	sessionId_ = v;
}

void WebApi::processQueue() {
	if (!queuedRequests_.size() || inProgressRequests_.size() >= 50) return;
	QueuedRequest& r = queuedRequests_.takeFirst();

	QString url = baseUrl_ + "/" + r.path;
	QUrlQuery query = r.query;

	if (sessionId_ != "") {
		query.addQueryItem("session", sessionId_);
	}

	url += "?" + query.toString(QUrl::FullyEncoded);

	QNetworkRequest* request = new QNetworkRequest(url);
	request->setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");

	QNetworkReply* reply = NULL;

	if (r.method == jop::PATCH) {
		// TODO: Delete buffer when done
		QBuffer* buffer = new QBuffer();
		buffer->open(QBuffer::ReadWrite);
		buffer->write(r.data.toString(QUrl::FullyEncoded).toUtf8());
		buffer->seek(0);
		r.buffer = buffer;
		reply = manager_.sendCustomRequest(*request, "PATCH", buffer);
	}

	if (r.method == jop::GET) {
	   reply = manager_.get(*request);
	}

	if (r.method == jop::POST) {
		reply = manager_.post(*request, r.data.toString(QUrl::FullyEncoded).toUtf8());
	}

	if (r.method == jop::PUT) {
		reply = manager_.put(*request, r.data.toString(QUrl::FullyEncoded).toUtf8());
	}

	if (r.method == jop::DEL) {
		reply = manager_.deleteResource(*request);
	}

	if (!reply) {
		qWarning() << "WebApi::processQueue(): reply object was not created - invalid request method";
		return;
	}

	r.reply = reply;
	r.request = request;
	connect(reply, SIGNAL(error(QNetworkReply::NetworkError)), this, SLOT(request_error(QNetworkReply::NetworkError)));

	QStringList cmd;
	cmd << "curl";
	if (r.method == jop::PUT) cmd << "-X" << "PUT";
	if (r.method == jop::PATCH) cmd << "-X" << "PATCH";
	if (r.method == jop::DEL) cmd << "-X" << "DELETE";

	if (r.method != jop::GET && r.method != jop::DEL) {
		cmd << "--data" << "'" + r.data.toString(QUrl::FullyEncoded) + "'";
	}
	cmd << "'" + url + "'";

	qDebug().noquote() << cmd.join(" ");

	inProgressRequests_.push_back(r);
}

void WebApi::request_finished(QNetworkReply *reply) {
	QByteArray responseBodyBA = reply->readAll();
	QJsonObject response;
	QJsonParseError err;
	QJsonDocument doc = QJsonDocument::fromJson(responseBodyBA, &err);
	if (err.error != QJsonParseError::NoError) {
		qWarning() << "Could not parse JSON:" << err.errorString();
		qWarning().noquote() << QString(responseBodyBA);
	} else {
		response = doc.object();
		if (response.contains("error") && !response["error"].isNull()) {
			qWarning().noquote() << "API error:" << QString(responseBodyBA);
		}
	}

	for (int i = 0; i < inProgressRequests_.size(); i++) {
		QueuedRequest r = inProgressRequests_[i];
		if (r.reply == reply) {
			inProgressRequests_.erase(inProgressRequests_.begin() + i);
			emit requestDone(response, r.tag);
			break;
		}
	}

	processQueue();
}

void WebApi::request_error(QNetworkReply::NetworkError e) {
	qDebug() << "Network error" << e;
}

File: ReactNativeClient/src/web-api.js
import { Log } from 'src/log.js';
import { stringify } from 'query-string';

class WebApi {

	constructor(baseUrl) {
		this.baseUrl_ = baseUrl;
		this.session_ = null;
	}

	setSession(v) {
		this.session_ = v;
	}

	session() {
		return this.session_;
	}

	makeRequest(method, path, query, data) {
		let url = this.baseUrl_;
		if (path) url += '/' + path;
		if (query) url += '?' + stringify(query);
		let options = {};
		options.method = method.toUpperCase();
		if (data) {
			let formData = null;
			if (method == 'POST') {
				formData = new FormData();
				for (var key in data) {
					if (!data.hasOwnProperty(key)) continue;
					formData.append(key, data[key]);
				}
			} else {
				options.headers = { 'Content-Type': 'application/x-www-form-urlencoded' };
				formData = stringify(data);
			}

			options.body = formData;
		}

		return {
			url: url,
			options: options
		};
	}

	static toCurl(r, data) {
		let o = r.options;
		let cmd = [];
		cmd.push('curl');
		if (o.method == 'PUT') cmd.push('-X PUT');
		if (o.method == 'PATCH') cmd.push('-X PATCH');
		if (o.method == 'DELETE') cmd.push('-X DELETE');
		if (o.method != 'GET' && o.method != 'DELETE') {
			cmd.push("--data '" + stringify(data) + "'");
		}
		cmd.push("'" + r.url + "'");
		return cmd.join(' ');
	}

	exec(method, path, query, data) {
		return new Promise((resolve, reject) => {
			if (this.session_) {
				query = query ? Object.assign({}, query) : {};
				if (!query.session) query.session = this.session_;
			}

			let r = this.makeRequest(method, path, query, data);

			Log.debug(WebApi.toCurl(r, data));

			fetch(r.url, r.options).then(function(response) {
				let responseClone = response.clone();
				return response.json().then(function(data) {
					if (data && data.error) {
						reject(new Error(data.error));
					} else {
						resolve(data);
					}
				}).catch(function(error) {
					responseClone.text().then(function(text) {
						reject(new Error('Cannot parse JSON: ' + text));
					});
				});
			}).then(function(data) {
				resolve(data);
			}).catch(function(error) {
				reject(error);
			});
		});
	}

	get(path, query) {
		return this.exec('GET', path, query);
	}

	post(path, query, data) {
		return this.exec('POST', path, query, data);
	}

	put(path, query, data) {
		return this.exec('PUT', path, query, data);
	}

	patch(path, query, data) {
		return this.exec('PATCH', path, query, data);
	}

	delete(path, query) {
		return this.exec('DELETE', path, query);
	}

}

export { WebApi };